tools/cldr-code/src/main/java/org/unicode/cldr/util/personname/PersonNameFormatter.java

```diff
@@ -0,0 +1,580 @@
+         */
+        public NamePatternData(ImmutableMap<ULocale, Order> localeToOrder, ImmutableListMultimap<ParameterMatcher, NamePattern> formatParametersToNamePattern) {
+            this.localeToOrder = localeToOrder == null ? ImmutableMap.of() : localeToOrder;
+            this.parameterMatcherToNamePattern = formatParametersToNamePattern;
+            // TODO: check formatParametersToNamePattern for validity
+            // * no null values
+            // * no ParameterMatcher should be completely masked by previous ones
+            // * each entry in ParameterMatcher must have at least one NamePattern
+            // * the final entry must have MATCH_ALL as the key
+        }
+
+        // TODO: add method that takes a string instead of a NamePatten
+
+        @Override
+        public String toString() {
+            return "{" + (localeToOrder.isEmpty() ? "" : "localeToOrder=" + localeToOrder + "\n\t\t")
+                + show(parameterMatcherToNamePattern) + "}";
+        }
+
+        private String show(ImmutableListMultimap<ParameterMatcher, NamePattern> multimap) {
+//            StringBuilder result = new StringBuilder();
+//            for (Entry<ParameterMatcher, Collection<NamePattern>> entry : multimap.asMap().entrySet()) {
+//                ParameterMatcher key = entry.getKey();
+//                if
+//            }
+            return parameterMatcherToNamePattern.asMap().toString();
+        }
+    }
+
+    public static interface NameObject {
+        public String getBestValue(ModifiedField modifiedField, Set<Modifier> unhandledModifiers);
+        public ULocale getNameLocale();
+        public Set<Field> getAvailableFields();
+    }
+
+    /**
+     * Simple implementation for testing and using in CLDR examples.
+     * Immutable
+     */
+    public static class SimpleNameObject implements NameObject {
+        private final ULocale nameLocale;
+        private final ImmutableMap<ModifiedField, String> patternData;
+
+        @Override
+        public Set<Field> getAvailableFields() {
+            Set<Field> result = EnumSet.noneOf(Field.class);
+            for (Entry<ModifiedField, String> entry : patternData.entrySet()) {
+                result.add(entry.getKey().getField());
+            }
+            return ImmutableSet.copyOf(result);
+        }
+
+        @Override
+        public String getBestValue(ModifiedField modifiedField, Set<Modifier> unhandledModifiers) {
+            // TODO flesh out the SimpleNameObject to handle modifiers
+            // and return the unhandled ones
+            // For now just return the item.
+            return patternData.get(modifiedField);
+        }
+
+        @Override
+        public ULocale getNameLocale() {
+            return nameLocale;
+        }
+
+        /**
+         * TODO Replace by builder
+         */
+        public SimpleNameObject(ULocale nameLocale, ImmutableMap<ModifiedField, String> patternData) {
```

Comment:
These are all for the internal CLDR implementation; and public just means public within CLDR. And it will be needed for more than testing, since we have to have an implementation for examples.

That being said, would be cleaner if this were in its own class file, so I can do that.